rasterio/sample.py

```diff
@@ -31,27 +41,29 @@ def sample_gen(dataset, xy, indexes=None, masked=False):
         those indexes.
 
     """
-    index = dataset.index
+    dt = dataset.transform
     read = dataset.read
+    height = dataset.height
+    width = dataset.width
 
     if indexes is None:
         indexes = dataset.indexes
     elif isinstance(indexes, int):
         indexes = [indexes]
 
-    for x, y in xy:
+    nodata = np.full(len(indexes), (dataset.nodata or 0),  dtype=dataset.dtypes[0])
+    if masked:
+        # Masks for masked arrays are inverted (False means valid)
+        mask = [MaskFlags.all_valid not in dataset.mask_flag_enums[i-1] for i in indexes]
+        nodata = np.ma.array(nodata, mask=mask)
 
-        row_off, col_off = index(x, y)
+    for pts in _grouper(xy, 256):
+        pts = zip(*filter(None, pts))
 
-        if row_off < 0 or col_off < 0 or row_off >= dataset.height or col_off >= dataset.width:
-            data = numpy.ones((len(indexes),), dtype=dataset.dtypes[0]) * (dataset.nodata or 0)
-            if masked:
-                mask = [False if MaskFlags.all_valid in dataset.mask_flag_enums[i - 1] else True for i in indexes]
-                yield numpy.ma.array(data, mask=mask)
+        for row_off, col_off in zip(*rowcol(dt, *pts)):
```

Comment:
@groutr what would you think about sorting these on row and column? My hypothesis is that the read() calls below would hit the raster block cache more often and we'd get another speed-up.

Reply:
Nevermind, we'll look at that later!